pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java

```diff
@@ -139,11 +144,22 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
         }
         break;
       case STRING:
-        for (String value : values) {
-          int dictId = dictionary.indexOf(value);
-          if (dictId >= 0) {
-            dictIdSet.add(dictId);
+        if (queryContext == null || values.size() < Integer.parseInt(queryContext.getQueryOptions()
+            .getOrDefault(CommonConstants.Broker.Request.QueryOptionKey.IN_PREDICATE_SORT_THRESHOLD,
+                CommonConstants.Broker.Request.QueryOptionValue.DEFAULT_IN_PREDICATE_SORT_THRESHOLD))) {
+          for (String value : values) {
+            int dictId = dictionary.indexOf(value);
+            if (dictId >= 0) {
+              dictIdSet.add(dictId);
+            }
           }
+        } else {
+          List<String> sortedValues =
+              queryContext.getOrComputeSharedValue(List.class, Equivalence.identity().wrap(inPredicate), v -> {
+                values.sort(String::compareTo);
```

Comment:
This can potentially cause problem as `values` is shared across segments, and this method is modifying it. We might want to make a clone of the list and sort on the clone.
(minor) This is equivalent to `values.sort(null)`